src/java_bytecode/java_object_factory.cpp

```diff
@@ -56,38 +172,81 @@ void gen_nondet_init(
         // make null
         null_pointer_exprt null_pointer_expr(pointer_type);
         code_assignt code(expr, null_pointer_expr);
+        code.add_source_location()=loc;
         init_code.copy_to_operands(code);
 
         return;
       }
     }
 
-    // build size expression
-    exprt object_size=size_of_expr(subtype, ns);
-
-    if(subtype.id()!=ID_empty && !object_size.is_nil())
-    {
-      // malloc expression
-      side_effect_exprt malloc_expr(ID_malloc);
-      malloc_expr.copy_to_operands(object_size);
-      malloc_expr.type()=pointer_type;
+    code_labelt set_null_label;
+    code_labelt init_done_label;
 
-      code_assignt code(expr, malloc_expr);
-      init_code.copy_to_operands(code);
+    static size_t synthetic_constructor_count=0;
 
-      // dereference expression
-      dereference_exprt deref_expr(expr, subtype);
+    if(!assume_non_null)
+    {
+      auto returns_null_sym=
+          new_tmp_symbol(symbol_table, "opaque_returns_null");
+      returns_null_sym.type=c_bool_typet(1);
+      auto returns_null=returns_null_sym.symbol_expr();
+      auto assign_returns_null=
+          code_assignt(returns_null, get_nondet_bool(returns_null_sym.type));
+      assign_returns_null.add_source_location()=loc;
+      init_code.move_to_operands(assign_returns_null);
+
+      auto set_null_inst=
+        code_assignt(expr, null_pointer_exprt(pointer_type));
+      set_null_inst.add_source_location()=loc;
+
+      std::ostringstream fresh_label_oss;
+      fresh_label_oss<<"post_synthetic_malloc_"
+          <<(++synthetic_constructor_count);
+      std::string fresh_label=fresh_label_oss.str();
+      set_null_label=code_labelt(fresh_label, set_null_inst);
+
+      init_done_label=code_labelt(fresh_label+"_init_done", code_skipt());
+
+      code_ifthenelset
+        null_check;
+      exprt null_return=constant_exprt("0", returns_null_sym.type);
```

Comment:
done, `message_handler` is now threaded downwards, `java_object_factory` inherits from `message`